udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala

```diff
@@ -45,19 +42,1834 @@ class OpcodeSuite extends FunSuite {
+  }
+
+  test("string test - charAt") {
+    val myudf: (String) => Int = a => {
+      a.charAt(2).toInt
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("mh7mmmm").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new", lit(55)) // utf8 for "7"
+    checkEquiv(result, ref)
+  }
+
+  test("string test - contains") {
+    val myudf: (String) => Boolean = a => {
+      a.contains("hmmm")
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("mh7mmmm","mhmmmm").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new", col("x").contains("hmmm"))
+    checkEquiv(result, ref)
+  }
+
+
+  test("string test - indexOf - case 1 - str,int") {
+    val myudf: (String) => Int = a => {
+      a.indexOf("c",1)
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("abcde").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new",lit(2))
+    checkEquiv(result, ref)
+  }
+
+
+  test("string test - indexOf - case 2 - char - single quotes,int") {
+    val myudf: (String) => Int = a => {
+      a.indexOf('c',1)
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("abcde").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new",lit(2))
+    checkEquiv(result, ref)
+  }
+
+  test("string test - indexOf - case 3 - char - utf value,int") {
+    val myudf: (String) => Int = a => {
+      a.indexOf(99,1)
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("abcde").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new",lit(2))
+    checkEquiv(result, ref)
+  }
+
+
+  test("string test - indexOf - case 4 - char - utf value") {
+    val myudf: (String) => Int = a => {
+      a.indexOf(99)
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("abcde").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new",lit(2))
+    checkEquiv(result, ref)
+  }
+
+
+  test("string test - indexOf - case 5 - char - single quotes") {
+    val myudf: (String) => Int = a => {
+      a.indexOf('c')
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("abcde").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new",lit(2))
+    checkEquiv(result, ref)
+  }
+
+
+  test("string test - indexOf - case 6 - str") {
+    val myudf: (String) => Int = a => {
+      a.indexOf("c")
+    }
+    val u = makeUdf(myudf)
+    val dataset = List("abcde").toDF("x").repartition(1)
+    val result = dataset.withColumn("new", u(col("x")))
+    val ref = dataset.withColumn("new",lit(2))
+    checkEquiv(result, ref)
+  }
+
+  // this test is an expected fallback to JVM execution due to no support for String.codePointAt
```

Comment:
We should add a mode to the compiler where if we are under tests, it will throw an exception when it can't replace. This would allow a test that the exception is thrown. For the unsupported case it's not as useful as for the supported case, where we would want to make sure that we didn't fallback automatically.

I think we can probably use this config: https://github.com/NVIDIA/spark-rapids/blob/branch-0.2/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala#L478, but we need to check for it in the udf compiler.

Reply:
Can we use the method here: https://github.com/NVIDIA/spark-rapids/pull/593/files#diff-96ccb4900fc50806e44aecb1039484afR67 ? 
like 
inside checkEquiv:
    if isUnderTest:
        isAllCompiled or say is AllReplaced.

Reply:
Yes this would work I think, as long as we only have 1 UDF to replace. Seems like all tests fall under this bucket now.

Reply:
I just tried this and there're 10 failed cases:

math functions - unsupported
long div and mod
loops - fallback test
string test - charAt
string test - indexOf - case 2 - char - single quotes,int
string test - indexOf - case 3 - char - utf value,int
string test - indexOf - case 4 - char - utf value
string test - indexOf - case 5 - char - single quotes
string test - codePointAt method
string test - matches method

Some of them are expected to fall back but  not sure about the others. @seanprime7 can you help to verify the execution plan?

Reply:
These are all expected to fall back except "long div and mod".
I will create a fix for "long div and mod" and push to this PR.

Reply:
Fixed it with 0abb51aef32d51286c0e4ee1bf7c1d69065018d0

Reply:
For the rest failed cases, do we need somehow make them pass? or leave them failed?

Reply:
Are you using `assert(udfIsCompiled(result))` to check if the udf is compiled successfully?
For the tests that fall back, you can use `assert(!udfIsCompiled(result))`

Reply:
Updated.